torch/distributed/_tensor/random.py

```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+
+    # override this function to define custom parallel random number generation
+    # to execute DTensor computation
+    @contextlib.contextmanager
+    def _parallel_region(self):
+        pass
+
+class DTensorRNG(ParallelRNGStyle):
+    # The default parallel random number generation style
+    def __init__(self):
+        super().__init__()
+
+    def _manual_seed(self, parallel_seed: int) -> None:
+        # do not pollute torch's default RNG state
+        device = torch.cuda.current_device()
+        with torch.random.fork_rng([device]):
+            torch.cuda.manual_seed(parallel_seed)
+            _rng_tracker.set_rng_state("parallel-rng", torch.cuda.get_rng_state())
+
+    @contextlib.contextmanager
+    def _parallel_region(self, spec: DTensorSpec):
+        # check if the parallel rng state has been synchronized or not
+        if not self.rng_state_is_sync("parallel-rng"):
+            raise RuntimeError(
+                "DTensorRNG requires random seed to be synchronized"
+                "before entering into a parallel region!"
+            )
+
+        # TODO: rewrite this set/restore logic using fork_rng()?
+        # fork rng state first since parallel RNG state should be used instead
+        old_state = torch.cuda.get_rng_state()  # a 16-byte ByteTensor
+        torch.cuda.set_rng_state(self.rng_states["parallel-rng"])
+        # set offset before entering the parallel region
+        mesh = spec.mesh
+        old_offset = _get_rng_offset(mesh)
+        set_pre_op_offset(spec)
+        try:
+            yield
+        finally:
+            # update offset to synchronize among ranks
+            set_post_op_offset(spec, old_offset)
+            _rng_tracker.set_rng_state("parallel-rng", torch.cuda.get_rng_state())
+            torch.cuda.set_rng_state(old_state)
+
+
+class MegatronRNG(ParallelRNGStyle):
+    def __init__(self):
+        super().__init__()
+
+    def _manual_seed(
+        self,
+        device_mesh: DeviceMesh,
+        default_seed: int = 1234,
+        tp_dim: int = 0,
+    ):
+        # Megatron maintains two sets of RNG states:
+        #   1. default state: torch's cuda RNG state. This is set to the same
+        #   across all ranks unless `data_parallel_random_init` is explicitly
+        #   set to True. Megatron uses this RNG state for non-tensor-parallel
+        #   regions.
+        #   2. tensor model parallel state: separate from the default state
+        #   and is only used in tensor-parallel regions. This is set to the same
+        #   across tensor parallel groups (i.e. along non-tp dimension) but different
+        #   within a tensor parallel group.
+
+        # calculate tensor_model_parallel_seed as Megatron does
+        # source: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/random.py
+        tensor_model_parallel_rank = device_mesh.get_coordinate()[tp_dim]
+        tensor_model_parallel_seed = default_seed + 2718 + tensor_model_parallel_rank
```

Comment:
I think this `2718` is a magic number here and if you want to reuse it can we make it a const instead of hard-code it in the code?

Reply:
It can be any positive number so the value itself won't make any impact on correctness only if this value is the same among all ranks but this is SPMD so no problem. That's why I don't think this value is "magical" enough to be made a constant variable.